### PR TITLE
refactor(rubygems): Use schema for fallback API

### PR DESCRIPTION
**Pull request by zharinov**
## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/datasource/rubygems/__fixtures__/rails/info.json [hunk 1]
```diff
@@ -1 +1 @@
...
...nvention over configuration.","licenses":["MIT"],"metadata":{},"sha":"d9ff5d9be16ee277dfc8f3c760bf171aa497d8685ec5c8988fba21a3dbd72cd5","project_uri":"https://rubygems.org/gems/rails","gem_uri":"https://rubygems.org/gems/rails-5.2.2.gem","homepage_uri":"http://rubyonrails.org","wiki_uri":"","documentation_uri":"http://api.rubyonrails.org","mailing_list_uri":"http://groups.google.com/group/rubyonrails-talk","source_code_uri":"http://github.com/rails/rails","bug_tracker_uri":"http://github.com/rails/rails/issues","changelog_uri":"https://www.railschangelog.com/","dependencies":{"development":[],"runtime":[{"name":"actioncable","requirements":"= 5.2.2"},{"name":"actionmailer","requirements":"= 5.2.2"},{"name":"actionpack","requirements":"= 5.2.2"},{"name":"actionview","requirements":"= 5.2.2"},{"name":"activejob","requirements":"= 5.2.2"},{"name":"activemodel","requirements":"= 5.2.2"},{"name":"activerecord","requirements":"= 5.2.2"},{"name":"activestorage","requirements":"= 5.2.2"},{"name":"activesupport","requirements":"= 5.2.2"},{"name":"bundler","requirements":"\u003e= 1.3.0"},{"name":"railties","requirements":"= 5.2.2"},{"name":"sprockets-rails","requirements":"\u003e= 2.0.0"}]}}
```
Comment (zharinov) [inline lib/modules/datasource/rubygems/__fixtures__/rails/info.json:1]: The only change here: make changelog URL non-null

**Comment by zharinov** [comment]
Worked fine locally

File: lib/modules/datasource/rubygems/get.ts [hunk 2]
```diff
@@ -1,185 +1,182 @@
...
     }
+
+    const packageName = config.packageName.toLowerCase();
+
     const hostname = parseUrl(registryUrl)?.hostname;
-    if (hostname && this.knownFallbackHosts.includes(hostname)) {
-      return this.getDependencyFallback(packageName, registryUrl);
-    }
-    return this.getDependency(packageName, registryUrl);
+    return hostname === 'rubygems.pkg.github.com' || hostname === 'gitlab.com'
```
[the remark this answers is not in the document]
Comment (zharinov) [reply lib/modules/datasource/rubygems/get.ts:78]: Too much premature abstraction in this file, maybe let's transform it to the array when there is at least 3 items?

File: lib/modules/datasource/rubygems/__snapshots__/index.spec.ts.snap [hunk 3]
```diff
@@ -3777,2380 +2422,1364 @@ exports[`modules/datasource/rubygems/index getReleases uses rubygems.org if no r
...
       "releaseTimestamp": "2018-11-28T22:55:23.827Z",
-      "rubyPlatform": "ruby",
-      "rubyVersion": ">= 2.2.2",
-      "rubygemsVersion": ">= 1.8.11",
       "version": "5.2.2.rc1",
     },
     {
       "releaseTimestamp": "2018-12-04T18:15:02.233Z",
-      "rubyPlatform": "ruby",
-      "rubyVersion": ">= 2.2.2",
```
[the remark this answers is not in the document]
Comment (zharinov) [reply lib/modules/datasource/rubygems/__snapshots__/index.spec.ts.snap:6152]: I can't find any usage of these fields, in fact there are no any types for these. Potentially, this is useful data that can be stored in `constraints` record, but as soon as `getPackageReleases` is stripping the `constraints` field, I gave up to parsing it.

File: lib/modules/datasource/rubygems/__fixtures__/rails/info.json (hunk 1 above)
Comment (zharinov) [reply lib/modules/datasource/rubygems/__fixtures__/rails/info.json:1]: It can, I've fixed it for coverage, I'll drop this huge file later anyway

File: lib/modules/datasource/rubygems/__snapshots__/index.spec.ts.snap (hunk 3 above)
[the remark this answers is not in the document]
Comment (viceice) [reply lib/modules/datasource/rubygems/__snapshots__/index.spec.ts.snap:6152]: please add as contraints, they get used here:
https://github.com/renovatebot/renovate/blob/df583b67995395e0be3701cbb42c8d89f11cbe7f/lib/modules/datasource/index.ts#L403-L427

[the remark this answers is not in the document]
Comment (zharinov) [reply lib/modules/datasource/rubygems/__snapshots__/index.spec.ts.snap:6152]: I'm not sure if it's useful, since this is the fallback implementation and the most of the packages won't have these `constraints` fields anyway.

[the remark this answers is not in the document]
Comment (viceice) [reply lib/modules/datasource/rubygems/__snapshots__/index.spec.ts.snap:6152]: @rarkins WDYT?

[the remark this answers is not in the document]
Comment (zharinov) [reply lib/modules/datasource/rubygems/__snapshots__/index.spec.ts.snap:6152]: Probably, we may want to prefer this API over `/versions` endpoint, since there are constraints and `releaseTimestamp` data available?

[the remark this answers is not in the document]
Comment (rarkins) [reply lib/modules/datasource/rubygems/__snapshots__/index.spec.ts.snap:6152]: I think we would overload rubygems.org if we used it in place of `/versions`.

At the minimum, we'd want to use it with essentially infinite cache time.

[the remark this answers is not in the document]
Comment (zharinov) [reply lib/modules/datasource/rubygems/__snapshots__/index.spec.ts.snap:6152]: Is 2 cached requests per package critical enough for rubygems?

[the remark this answers is not in the document]
Comment (zharinov) [reply lib/modules/datasource/rubygems/__snapshots__/index.spec.ts.snap:6152]: I've noted #20094 removes `/versions` fetching and prefers to query API for the exact same reason, in order to address #3670

[the remark this answers is not in the document]
Comment (zharinov) [reply lib/modules/datasource/rubygems/__snapshots__/index.spec.ts.snap:6152]: So, I think we need to use both of the APIs using different cache strategies

[the remark this answers is not in the document]
Comment (rarkins) [reply lib/modules/datasource/rubygems/__snapshots__/index.spec.ts.snap:6152]: We should not replace /versions was the primary approach. The API can be teed to (heavily) cache metadata

[the remark this answers is not in the document]
Comment (zharinov) [reply lib/modules/datasource/rubygems/__snapshots__/index.spec.ts.snap:6152]: I understand, so [this comment](https://github.com/renovatebot/renovate/pull/20094#pullrequestreview-1292892214) still makes sense, right?

[the remark this answers is not in the document]
Comment (zharinov) [reply lib/modules/datasource/rubygems/__snapshots__/index.spec.ts.snap:6152]: Regarding constraints, are `rubyPlatform`, `rubyVersion` and `rubygemsVersion` valid fields for this?

[the remark this answers is not in the document]
Comment (zharinov) [reply lib/modules/datasource/rubygems/__snapshots__/index.spec.ts.snap:6152]: Actually, seems like it should be: `ruby`, `rubygems` and... don't know how to better describe `platform` field, probably just drop it?

[the remark this answers is not in the document]
Comment (zharinov) [reply lib/modules/datasource/rubygems/__snapshots__/index.spec.ts.snap:6152]: UPD: updated to `ruby`, `rubygems` and `platform`

File: lib/util/http/index.ts [hunk 4]
```diff
@@ -221,21 +221,14 @@ export class Http<Opts extends HttpOptions = HttpOptions> {
...
-    });
-  }
-
   getBuffer(
     url: string,
     options: HttpOptions = {}
-  ): Promise<HttpResponse<Buffer> | null> {
-    return this.requestBuffer(url, options);
+  ): Promise<HttpResponse<Buffer>> {
+    return this.request<Buffer>(url, {
```
Comment (viceice) [inline lib/util/http/index.ts:228]: unrelated change?

**Review by viceice — COMMENTED** [review]
otherwise LGTM

**Review by viceice — APPROVED** [review]
